internal/component/mimir/rules/kubernetes/rules.go

```diff
@@ -146,7 +159,60 @@ func New(o component.Options, args Arguments) (*Component, error) {
 	return c, nil
 }
 
+func newNoInit(o component.Options, args Arguments) (*Component, error) {
+	m := newMetrics()
+	m.mustRegister(o.Registerer)
+
+	clusterSvc, err := o.GetServiceData(cluster.ServiceName)
+	if err != nil {
+		return nil, fmt.Errorf("getting cluster service failed: %w", err)
+	}
+
+	c := &Component{
+		log:            o.Logger,
+		opts:           o,
+		args:           args,
+		leader:         newComponentLeadership(o.ID, o.Logger, clusterSvc.(cluster.Cluster)),
+		configUpdates:  make(chan ConfigUpdate),
+		clusterUpdates: make(chan struct{}),
+		ticker:         time.NewTicker(args.SyncInterval),
+		metrics:        m,
+	}
+
+	return c, nil
+}
+
 func (c *Component) Run(ctx context.Context) error {
+	c.startupWithRetries(ctx, c.leader, c, c)
+
+	for {
+		// iteration only returns a sentinel error to indicate shutdown, otherwise it handles
+		// any errors encountered itself by logging and marking the component as unhealthy.
+		if err := c.iteration(ctx, c.leader, c, c); err != nil {
+			break
+		}
+	}
+
+	return nil
+}
+
+func (c *Component) Update(newConfig component.Arguments) error {
+	errChan := make(chan error)
+	c.configUpdates <- ConfigUpdate{
+		args: newConfig.(Arguments),
+		err:  errChan,
+	}
+	return <-errChan
+}
```

Comment:
We don't make any guarantees that Update never gets called before Run, so I'm worried about the possibility of a future deadlock here if the controller gets updated. 

Since we set the component health on an error anyway, maybe we don't need to send an error back to the Update method?

Reply:
Sure, I can change that. This was preserving the existing behavior of the component.

Reply:
Thanks for pointing that out. I think it's OK to leave the other issue that was preserving behavior, but since this is a smaller change and I'm worried about it suddenly breaking I'd like to fix at least this one up.

Reply:
Fixed in de9737a21629cbc064f594ef0808125d06dfdb52